src/Security/MemberAuthenticator/CookieAuthenticationHandler.php

```diff
@@ -0,0 +1,238 @@
+            } else {
+                // Check for expired token
+                $expiryDate = new \DateTime($rememberLoginHash->ExpiryDate);
+                $now = DBDatetime::now();
+                $now = new \DateTime($now->Rfc2822());
+                if ($now > $expiryDate) {
+                    $member = null;
+                }
+            }
+        }
+
+        if ($member) {
+            if ($this->cascadeLogInTo) {
+                // @todo look at how to block "regular login" triggers from happening here
+                // @todo deal with the fact that the Session::current_session() isn't correct here :-/
+                $this->cascadeLogInTo->logIn($member, false, $request);
+                //\SilverStripe\Dev\Debug::message('here');
+            }
+
+            // @todo Consider whether response should be part of logIn() as well
+
+            // Renew the token
+            if ($rememberLoginHash) {
+                $rememberLoginHash->renew();
+                $tokenExpiryDays = RememberLoginHash::config()->uninherited('token_expiry_days');
+                Cookie::set(
+                    $this->getTokenCookieName(),
+                    $member->ID . ':' . $rememberLoginHash->getToken(),
+                    $tokenExpiryDays,
+                    null,
+                    null,
+                    false,
+                    true
+                );
+            }
+
+            return $member;
+
+            // Audit logging hook
+            $member->extend('memberAutoLoggedIn');
+        }
+    }
+
+    /**
+     * @inherit
+     */
+    public function logIn(Member $member, $persistent, HTTPRequest $request)
+    {
+        // @todo couple the cookies to the response object
+
+        // Cleans up any potential previous hash for this member on this device
+        if ($alcDevice = Cookie::get($this->getDeviceCookieName())) {
+            RememberLoginHash::get()->filter('DeviceID', $alcDevice)->removeAll();
+        }
+
+        // Set a cookie for persistent log-ins
+        if ($persistent) {
+            $rememberLoginHash = RememberLoginHash::generate($member);
+            $tokenExpiryDays = RememberLoginHash::config()->uninherited('token_expiry_days');
+            $deviceExpiryDays = RememberLoginHash::config()->uninherited('device_expiry_days');
+            Cookie::set(
+                $this->getTokenCookieName(),
+                $member->ID . ':' . $rememberLoginHash->getToken(),
+                $tokenExpiryDays,
+                null,
+                null,
+                null,
+                true
+            );
+            Cookie::set(
+                $this->getDeviceCookieName(),
+                $rememberLoginHash->DeviceID,
+                $deviceExpiryDays,
+                null,
+                null,
+                null,
+                true
+            );
+
+        // Clear a cookie for non-persistent log-ins
+        } else {
+            $this->logOut($request);
```

Comment:
I'm going to split out the cleanup of the cookies. That makes the most sense to me.